File: dynamic programming/Rob2.cs
public class Solution {
    public int Rob(int[] nums) {
        // special cases
        if(nums.Length == 0) return 0;
        if(nums.Length == 1) return nums[0];

        int previousRobbed = 0, currentRobbed = 0, 
        previousNotRobbed = 0, currentNotRobbed = 0;
        // we will excluse the last house
        for(int i=0; i<nums.Length-1; i++){
            int tempNotRobbed = previousNotRobbed;
            // set previous to current
            previousNotRobbed = currentNotRobbed;
            // set current to temp
            currentNotRobbed = Math.Max(tempNotRobbed + nums[i], previousNotRobbed);
            // same for robbed
            int tempRobbed = previousRobbed;
            // set previous to current
            previousRobbed = currentRobbed;
            // set current to temp
            currentRobbed = Math.Max(tempRobbed + nums[i+1], previousRobbed);   
        }
        // since there are two possibilities of shifting
        // we return the max value between currentRobbed and currentNotRobbed
        return Math.Max(currentRobbed, currentNotRobbed);
    }
}

File: dynamic programming/Rob2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        if len(nums) == 1:
            return nums[0]
        previousRobbed, currentRobbed, previousNotRobbed, currentNotRobbed = 0, 0, 0, 0
        
        for i in range(len(nums)-1):
            tempNotRobbed = previousNotRobbed
            previousNotRobbed = currentNotRobbed
            currentNotRobbed = max(tempNotRobbed + nums[i], previousNotRobbed)
            
            tempRobbed = previousRobbed
            previousRobbed = currentRobbed
            currentRobbed = max(tempRobbed + nums[i+1], previousRobbed)
        return max(currentRobbed, currentNotRobbed)